### ISSUE TITLE
Storage interface should be bytes, not strings

### ISSUE THREAD

**Issue by austinabell** [near/near-sdk-js#90]
Forcing string usage for all storage interactions like storage read, write, register reads is inefficient because it requires data to be stringified in some way (JSON, base64) which is very space inefficient and won't scale well.

Low-level options, that pass slices of bytes should be provided to enable people to use things like borsh or some binary encoding protocol that's more efficient.

The string APIs can still be built on top of these for convenience, but it seems more useful to provide a higher-level API that serializes objects as JSON, for example

**Comment by ailisp** [comment]
Your points are correct and I agree. But due to quickjs C API limitation, what you seen "String"s here are actually bytes (!), mentioned here: https://github.com/near/near-sdk-js#about-type

It isn't a very obvious in the documentation, and we probably can document it in a better way, such as mark these as "Bytes", although internally, it still stores as quickjs string object.

**Comment by austinabell** [comment]
But this isn't necessarily a good thing, shouldn't the interface not express that internal quickjs limitation and also so that if we wanted to swap the engine for the SDK we could without breaking the interface?

**Comment by ailisp** [comment]
In theory, we should provide an API that hide the internal representation, say, `bytes()`. So user would call all the apis like so:
```
jsvmStorageRead(bytes('key'))
// or
jsvmStorageRead(bytes(new Uint8Array([1, 2, 3]))
```

WDYT?

**Comment by austinabell** [comment]
> In theory, we should provide an API that hide the internal representation, say, `bytes()`. So user would call all the apis like so:
> 
> ```
> jsvmStorageRead(bytes('key'))
> // or
> jsvmStorageRead(bytes(new Uint8Array([1, 2, 3]))
> ```
> 
> WDYT?

Not sure how this fits into JS APIs, if this is some function as you show or some other class, but that is a valid path. Unsure if that is better since it's more verbose. I wonder if another option is to handle the possible inputs with something like `typeof`.

TLDR: I don't know, I'd get input from someone else with more exp with designing JS interfaces
